src/runtime/rpc/minrpc/minrpc_server.h

```
@@ -0,0 +1,598 @@
+          case RPCCode::kShutdown: {
+            this->Shutdown();
+            return;
+          }
+          default: {
+            this->ThrowError(RPCServerStatus::kUnknownRPCCode);
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  void Shutdown() {
+    arena_.FreeAll();
+    io_.Close();
+  }
+
+  void HandleNormalCallFunc() {
+    uint64_t call_handle;
+    TVMValue* values;
+    int* tcodes;
+    int num_args;
+    TVMValue ret_value[3];
+    int ret_tcode[3];
+
+    this->Read(&call_handle);
+    RecvPackedSeq(&values, &tcodes, &num_args);
+
+    int call_ecode = TVMFuncCall(
+        reinterpret_cast<void*>(call_handle),
+        values, tcodes, num_args,
+        &(ret_value[1]), &(ret_tcode[1]));
+
+    if (call_ecode == 0) {
+      // Return value encoding as in LocalSession
+      int rv_tcode = ret_tcode[1];
+      ret_tcode[0] = kDLInt;
+      ret_value[0].v_int64 = rv_tcode;
+      if (rv_tcode == kTVMNDArrayHandle) {
+        ret_tcode[1] = kTVMDLTensorHandle;
+        ret_value[2].v_handle = ret_value[1].v_handle;
+        ret_tcode[2] = kTVMOpaqueHandle;
+        this->ReturnPackedSeq(ret_value, ret_tcode, 3);
+      } else if (rv_tcode == kTVMPackedFuncHandle ||
+                 rv_tcode == kTVMModuleHandle) {
+        ret_tcode[1] = kTVMOpaqueHandle;
+        this->ReturnPackedSeq(ret_value, ret_tcode, 2);
+      } else {
+        this->ReturnPackedSeq(ret_value, ret_tcode, 2);
+      }
+    } else {
+      this->ReturnLastTVMError();
+    }
+  }
+
+  void HandleCopyFromRemote() {
+    uint64_t handle, offset, num_bytes;
+    TVMContext ctx;
+    DLDataType type_hint;
+
+    this->Read(&handle);
+    this->Read(&offset);
+    this->Read(&num_bytes);
+    this->Read(&ctx);
+    this->Read(&type_hint);
+
+    char* data_ptr;
+    int call_ecode = 0;
+    if (ctx.device_type == kDLCPU) {
+      data_ptr = reinterpret_cast<char*>(handle) + offset;
+    } else {
+      data_ptr = this->ArenaAlloc<char>(num_bytes);
```

CHECK or return an error if this fails?

---

er, we should return an error and the reason i'm raising this is i'm wondering which one

---

also, s/char/uint8_t/?

---

Arena itself will call allocator's allocate, which will call IOHandler.ThrowError